### ISSUE TITLE
Strings: % formatting

### ISSUE THREAD

**Issue by nicolaslauquin** [SwiftGen/SwiftGen#577]
Hello,

thanks again for the great work with this tool .

**TLDR**: I want to display a string with a basic % in it, and it needs to be doubled %% with SwiftGen due to the internal usage of **String(format:** but my string don't has any parameter and one % should be enough for a basic string. 

**Context**
Version 6.0.2 
Template  structured-swift4.stencil

I have several strings in my app with some pourcentage *%* (ex of string 100% of success blablabla) and the issue with SwiftGen is it requires a double **%%* but as I'm using an external service to store and handle my translations on this multi-platform project we have to do something specific, and it should not. The service is working with an universal placeholder [%] that is translated to one or two **%** depending of the plateform and context. On Android it is always one %, on iOS it is one `%` if no parameter and two **%%** if at least one parameter provided (such as %@).

It makes sense as it is the way iOS is working with string manipulation, depending of usage of format&parameters. After thinking to several solutions, I decided to fix it at the source with a custom template. Here is the action I did:
 
**BEFORE**
```swift
private static func tr(_ table: String, _ key: String, _ args: CVarArg...) -> String {
  
    let format = NSLocalizedString(key, tableName: table, bundle: Bundle(for: BundleToken.self), comment: "")
    return String(format: format, locale: Locale.current, arguments: args)
 }
```

**AFTER**
```swift
private static func tr(_ table: String, _ key: String, _ args: CVarArg...) -> String {
  
  let format = NSLocalizedString(key, tableName: table, bundle: Bundle(for: BundleToken.self), comment: "")
  if args.count > 0 {
    return String(format: format, locale: Locale.current, arguments: args)
  } else {
    return format
  }
}
```

As the support told me I am not the only one contacting them for that subject with SwiftGen, I think it should be done for everyone but still want to exchange on this with you.

Thank for your time!

**Comment by djbe** [comment]
Hi @nicolaslauquin,

We've already had quite a few issues discussing this. Initially we added something like you have in #503 (issue #501), but we had to roll it back in #542. Users expect `%%` to get unescaped for them, see https://github.com/SwiftGen/SwiftGen/issues/501#issuecomment-427085029.

So you could use your custom template, but we can't accept such a change for the default template. If you're using the same translations for both iOS and Android, we recommend you use a service or tool that can handle such platform differences.

**Comment by AliSoftware** [comment]
Given that this question have been asked multiple times, it could be a good thing to write a note (markdown in [Documentation/Articles](https://github.com/SwiftGen/SwiftGen/tree/master/Documentation/Articles), with link from the README?) to explain that people need to use `%%`, why it can't be solved for all cases with the `if args.isEmpty` check and the problem / side effect it would bring, and links to past issues on this…

**Comment by AliSoftware** [comment]
@nicolaslauquin Just to give you an example of the problems we had (TL;DR for the other issues @djbe mentioned, which explain the problems in details):

>  On Android it is always one %, on iOS it is one % if no parameter and two %% if at least one parameter provided (such as %@).

So how does the service handle translations like "100% successful"? Does it consider it as a literal that shouldn't need to call `String(format: )`, as a human might thing at first when reading this? But given `% s` is a totally valid placeholder according to the `sprintf` spec (it's like `%s` but with a ` ` space padding character), according to the RFC we see a placeholder in that string so we can only assume you wanted to parametrize that translation with a `%s` string…

Same of course would arise for `0% data loss` which might look like a literal with no parameters at first but we would have no way of distinguish between that and a string with a `%d` placeholder with a space for padding…

**Comment by nicolaslauquin** [comment]
@AliSoftware @djbe Thanks all for this feeback. But still the discussion is useful for this specific usage.
I don't think the debate is really on lokalise it is more on `basic String` to display vs `String(format` usage. Should I respect the simple % for a basic string not transformed, and %% for a string that need to be formated with params.

> So how does the service handle translations like "100% successful"? Does it consider it as a literal that shouldn't need to call String(format: ), as a human might thing at first when reading this?
In such case, the string is on the tool declared as "100[%] successful" to respect each plateforms and be able to convert it. 
- `"100[%] successful"` is translated -> `"100% successful"' as it is a basic string without other universal placeholders than a [%] and so it don't need to be formatted.
- `"100[%] successful [%@]"` would be translated has `"100%% successful %@"` for swift to be compilant with `String(format: )`.
- FYI an easy to read universal placeholders description (https://docs.lokalise.co/developer-docs/universal-placeholders). We are using limited and *restricted* placeholders

> But given % s is a totally valid placeholder according to the sprintf spec (it's like %s but with a space padding character), according to the RFC we see a placeholder in that string so we can only assume you wanted to parametrize that translation with a %s string…

Ok, in such case `% s` will have 1 arg so won't be `args.isEmpty`. So not sur if the template is breaking something. Maybe I'm missing something?


> If you're using the same translations for both iOS and Android, we recommend you use a service or tool that can handle such platform differences.

Initially I was thinking like you and wanted to ask Lokalise to provide a specific option to always return %% which would works fine with SwiftGen, but after taking time thinking about this and other user usage, it is because SwiftGen is doing an exclusive and systematic usage of `String(Format:` we need to do so. So it seems to be an architecture decision that might impact other users to having two %% displayed in their UI and need to update the string. So I decided to modify the template instead of asking a feature to the service for a scenario which seems plateform compliant. (If I'm not missing something of course)

Can't wait to have your point of view about this.

**Comment by nicolaslauquin** [comment]
@AliSoftware yes, I read the long exchanges, it is true that the ```% d```  is complexifying the detection and would brake this scenario except if using a ```Non-breaking space``` instead of a ```space```. Here is a quick test:

<img width="389" alt="result" src="https://user-images.githubusercontent.com/566880/51383446-bcf7ce80-1b19-11e9-921b-70fccf653563.png">

We can keep the things here if extensively discussed, but a standard project is not compatible with SwiftGen and users need to update their text with % to %% if not format. This is not a big deal, majority of people can fix it quickly in the project, creating as specific template, or running a script. Just wanted to share my point of view.

**Comment by nicolaslauquin** [comment]
@AliSoftware I agree with point 1 and 3, but just I disagree on point 2 that Lokalise is generating a custom format. It is generating a `Localizable.strings` that is compliant with the expectation of the platform (and I would generate it the exact same way without the service). This is also why I wast some time trying to exchange with you and maybe found something that might work.
 
The difference is that a human knows if a basic string neesd to be consumed with parameter or not and so to choose between ```String``` or ```String(format:```. So on a standard iOS project I would need to migrate all ```%``` to ```%%``` to be able to use SwiftGen, which is using as default ```String(format:```.

Thank a lot for taking time to read and response! Keep the good work.

**Comment by nicolaslauquin** [comment]
@AliSoftware It was just about this specific part: `Can't drive us to change SwiftGen internal behaviour *just for people using lokalize*`. It is not only for Lokalize people, but also for people switching to SwiftGen.

For point 3, I think it is a bit complexe for the usage, even if it is working and yes it is also impacting the other plateforms. This idea could be the base of other fields of improvment with template param or placeholder to be able to address some specific usage.

**Comment by AliSoftware** [comment]
@nicolaslauquin Getting back at this: now that I had the occasion to discover Lokalise myself, I saw this option when downloading the translations:

![image](https://user-images.githubusercontent.com/216089/62052944-7e507880-b20e-11e9-8213-9282ac7e122d.png)

I think if you check the "Convert all [%] to %%" checkbox, it's supposed to generate `%%` for any literal percent... even if there's not any other `%d/%@/...` placeholder in the rest of the string. Which means they will be expected to always go thru `String(format:)`, so if you check that box I'm thinking it should solve the problem? (Haven't tried it yet since it's not my top priority right away, but if you have time to check it...). – PS: Apparently the corresponding flag for the CLI is `--escape_percent 1`

**Comment by grigorye** [comment]
Just in case, employing non-breakable space after `%` (instead of regular space character) may be a workaround that does not require `%%`/could keep strings compatible with other platforms.
